File: backend/app/schemas/competitors.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, HttpUrl, model_validator


class CompetitorAnalyzeRequest(BaseModel):
    seller_url: HttpUrl
    include_active: bool = True
    include_sold: bool = True
    limit: int = Field(default=50, ge=1, le=100)

    @model_validator(mode="after")
    def validate_status_selection(self):
        if not self.include_active and not self.include_sold:
            raise ValueError("at least one of include_active or include_sold must be true")
        return self


class CompetitorKeywordCreate(BaseModel):
    model_config = ConfigDict(protected_namespaces=())

    keyword: str
    category: str | None = "Competitor Research"
    brand: str | None = None
    model_group: str | None = None
    priority: int = Field(default=80, ge=1, le=999)
    active: bool = True


class CompetitorKeywordBulkCreate(BaseModel):
    model_config = ConfigDict(protected_namespaces=())

    keywords: list[str] | None = None
    category: str | None = "Competitor Research"
    brand: str | None = None
    model_group: str | None = None
    priority: int = Field(default=80, ge=1, le=999)
    active: bool = True


class CompetitorItemRead(BaseModel):
    id: int
    seller_id: int
    marketplace: str
    external_item_id: str
    title: str
    normalized_title: str | None = None
    item_url: str
    image_url: str | None = None
    price: float | None = None
    currency: str | None = None
    item_status: str
    source_url: str
    first_seen_at: datetime
    last_seen_at: datetime

    model_config = ConfigDict(from_attributes=True)


class CompetitorSellerRead(BaseModel):
    id: int
    marketplace: str
    seller_username: str
    seller_url: str
    fetch_status: str
    last_error: str | None = None
    active_count: int
    sold_count: int
    avg_active_price: float | None = None
    avg_sold_price: float | None = None
    last_analyzed_at: datetime | None = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class CompetitorAnalyzeResponse(BaseModel):
    seller: CompetitorSellerRead
    items: list[CompetitorItemRead]


class CompetitorKeywordSuggestion(BaseModel):
    keyword: str
    count: int


class CompetitorTopTerm(BaseModel):
    term: str
    count: int


class CompetitorInsights(BaseModel):
    seller_id: int
    seller_username: str
    active_count: int
    sold_count: int
    sell_through_rate: float | None = None
    avg_active_price: float | None = None
    avg_sold_price: float | None = None
    sold_active_price_gap: float | None = None
    top_sold_terms: list[CompetitorTopTerm]
    suggested_keywords: list[CompetitorKeywordSuggestion]

File: frontend/src/types/competitors.ts
export type CompetitorSeller = {
  id: number
  marketplace: string
  seller_username: string
  seller_url: string
  fetch_status: string
  last_error: string | null
  active_count: number
  sold_count: number
  avg_active_price: number | null
  avg_sold_price: number | null
  last_analyzed_at: string | null
  created_at: string
  updated_at: string
}

export type CompetitorItem = {
  id: number
  seller_id: number
  marketplace: string
  external_item_id: string
  title: string
  normalized_title: string | null
  item_url: string
  image_url: string | null
  price: number | null
  currency: string | null
  item_status: 'active' | 'sold' | string
  source_url: string
  first_seen_at: string
  last_seen_at: string
}

export type CompetitorAnalyzeResponse = {
  seller: CompetitorSeller
  items: CompetitorItem[]
}


export type CompetitorTopTerm = {
  term: string
  count: number
}

export type CompetitorKeywordSuggestion = {
  keyword: string
  count: number
}

export type CompetitorInsights = {
  seller_id: number
  seller_username: string
  active_count: number
  sold_count: number
  sell_through_rate: number | null
  avg_active_price: number | null
  avg_sold_price: number | null
  sold_active_price_gap: number | null
  top_sold_terms: CompetitorTopTerm[]
  suggested_keywords: CompetitorKeywordSuggestion[]
}
